p2p/exchange.go

```diff
@@ -105,56 +109,61 @@ func (ex *Exchange[H]) Stop(context.Context) error {
 	return nil
 }
 
-// Head requests the latest Header. Note that the Header must be verified thereafter.
-// NOTE:
-// It is fine to continue handling head request if the timeout will be reached.
-// As we are requesting head from multiple trusted peers,
-// we may already have some headers when the timeout will be reached.
+// Head requests the latest Header from trusted peers.
+//
+// The Head must be verified thereafter where possible.
+// We request
 func (ex *Exchange[H]) Head(ctx context.Context) (H, error) {
 	log.Debug("requesting head")
-	// create request
-	req := &p2p_pb.HeaderRequest{
-		Data:   &p2p_pb.HeaderRequest_Origin{Origin: uint64(0)},
-		Amount: 1,
+
+	reqCtx := ctx
+	if deadline, ok := ctx.Deadline(); ok {
+		// allocate 90% of callers set deadline for requests
+		// and give leftover to determine the bestHead from gathered responses
+		// this avoids DeadlineExceeded error when any of the peers are unresponsive
+		now := time.Now()
+		sub := deadline.Sub(now) * 9 / 10
+		var cancel context.CancelFunc
+		reqCtx, cancel = context.WithDeadline(ctx, now.Add(sub))
+		defer cancel()
```

Comment:
This solution has one downside. If at least one peer is unresponsive, we will have to wait for all 90% of the time. However, that's an edge case, and usually, peers would either respond or error out. After a lengthy investigation, it becomes clear that it's not worth optimizing for this edge case as it increases the complexity of the solution